### DISCUSSION TITLE
Memory leak on javaagent 1.5.1

### DISCUSSION THREAD

**Discussion by lmssy** [open-telemetry/opentelemetry-java-instrumentation#3960] [Q&A]
Here is an OOM dump below:
![image](https://user-images.githubusercontent.com/17894328/130907529-97239267-61f4-4af5-bd05-a00e9082130c.png)

![image](https://user-images.githubusercontent.com/17894328/130908158-42cccff4-b7a3-4127-9043-31d29a8bbd5e.png)

javaagent version: 1.5.1
params: -Dotel.metrics.exporter=none -Dotel.exporter.otlp.traces.endpoint=http://myip:4317 -Dotel.traces.sampler=parentbased_traceidratio -Dotel.traces.sampler.arg=1 -Dio.opentelemetry.context.enableStrictContext=true

I have applications run on 1.3.1 and 1.4.1 for a long time without OOM, yeaterday I upgrade to 1.5.1 and my application crashed by OOM.

Please have a look, thanks.

**Comment by anuraaga** [comment]
@lmssy I don't know if it will have better or worse performance, but can you try `-Dotel.metrics.exporter=logging` (or just leave it at default and accept that metric export requests are attempted but fail)?

@jsuereth @jkwatson I have a feeling we have an issue with the metrics SDK storage that it can grow forever if no exporter is configured to trigger resets.

**Reply by lmssy** [reply]
Ok, I will try -Dotel.metrics.exporter=logging to see if it happens again. thx.

**Reply by anuraaga** [reply]
If the flag change helps, it would confirm my hypothesis. In the meantime I have filed https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3962 and https://github.com/open-telemetry/opentelemetry-java/pull/3555

**Reply by lmssy** [reply]
Unfortunately memory leak still here after using -Dotel.metrics.exporter=logging.

![image](https://user-images.githubusercontent.com/17894328/131095939-8e267aae-a935-4020-b632-640cb9534eb8.png)

Details for Suspect1
![image](https://user-images.githubusercontent.com/17894328/131095530-de1dd53e-d88e-45f6-a2c2-96ca8f5430eb.png)

Details for Suspect2
![image](https://user-images.githubusercontent.com/17894328/131095483-bbf615ba-e6cb-4515-9872-8be7d319cc0b.png)

**Reply by jkwatson** [reply]
@jsuereth  are we keeping points around rather than summing them up in the buckets?

**Reply by jkwatson** [reply]
Now I'm confused. We only generate these during a data collection, which only happens if an exporter is configured. So, I don't understand how these are being kept around by anything, unless there's some exporter that is holding on to them indefinitely and not actually exporting them.

**Reply by jkwatson** [reply]
and, yes, there is an IntervalMetricReader wired up here, I see that above.

**Reply by jkwatson** [reply]
@lmssy how long are you waiting before grabbing the heap? The IntervalMetricReader has a default interval of 1 minute, so you would only see these getting swept out and GC'd on a minute interval, unless you explicitly configure a shorter time interval with the `otel.imr.export.interval` setting.

**Reply by jkwatson** [reply]
Also, now that 1.5.2 has been released, can you check this again, removing the logging exporter configuration altogether?
